Fixed #14094 -- Added support for unlimited models.CharField on PostgreSQL.

Elkasitu opened the pull request:
This commit introduced a new database feature flag
`supports_unlimited_varchar` which denotes which database backends
support defining `VARCHAR` columns with no/unlimited size.

When this feature flag is enabled for a given backend, CharField
definitions may rely on the default `max_length` value of `None` instead
of having to explicitly define one. It is however still possible to
explicitly pass a `max_length` and this will be enforced at the database
level.

This commit also implemented the necessary changes for the feature to
apply to the postgres backend.

carltongibson commented:
Just a query, on the ticket there was [a mention of `inspectdb`'s output here]https://code.djangoproject.com/ticket/14094#comment:3). Did you happen to look into that at all?

Elkasitu commented:
@carltongibson I didn't, I wasn't aware of this Django feature until just now nor what Joe meant by "much smoother" (faster? less code complexity? etc.).

I have just tested it in a dummy Django project and the result is incorrect Django code (`max_length=-1`), I will take a look at it later today and see what I can do.

Thanks for the heads up!

Elkasitu commented:
@carltongibson it should work now, I am not exactly happy with the implementation (it assumes that a negative `internal_size` for a `varchar` column means that the `varchar` is of unlimited size) but I'm not sure that there's a better alternative given the way that `inspectdb` sets field attributes.

Elkasitu commented:
I could define a constant in the backend's instropection class e.g. `UNLIMITED_VARCHAR_ISIZE = -1` (for pg) and use that in `inspectdb` to determine whether it is in fact a `varchar` of unlimited size

carltongibson commented on django/db/models/fields/__init__.py:
```diff
@@ -794,7 +794,11 @@ def db_type(self, connection):
         data = self.db_type_parameters(connection)
         try:
-            return connection.data_types[self.get_internal_type()] % data
+            column_type = connection.data_types[self.get_internal_type()]
+            # column_type is either a single-parameter function or a string
+            if callable(column_type):
+                return column_type(data)
+            return column_type % data
```
So I guess this comes down to whether to allow callables here, vs (say) a custom field (in `contrib.postgres`?) 

@charettes @felixxm — can I ask your opinion there please?

charettes replied:
I don't think that returning a `callable` is a huge deal but it can be avoided by having `DatabaseWrapper.data_types` entries turned instances of a `str` subclass that implements `__mod__`

```python
class Varchar(str):
    def __new__(cls):
        return super().__new__(cls, "varchar(%(max_length)s)")

    def __mod__(self, data):
        if data["max_length"] is None:
            return "varchar"
        return super().__mod__(data)
```

```python
>>> vc = Varchar()
'varchar(%(max_length))'
>>> isinstance(vc, str)
True
>>> vc % {'max_length': 255}
'varchar(355)'
>>> vc % {'max_length': None}
'varchar'

Elkasitu replied:
That's fair enough, I guess it depends on whether we esteem that there might be more cases like this in the future or not -- if it is the case I would prefer callables, otherwise Simon's suggestion is fine for a one-time thing.

Elkasitu commented:
Rebased on top of the changes introduced by #16368, some tests are failing but the test that fails seems unrelated to my changes

felixxm commented on django/core/management/commands/inspectdb.py:
```diff
@@ -329,7 +329,15 @@ def get_field_type(self, connection, table_name, row):
-            field_params["max_length"] = int(row.display_size)
+            if (size := int(row.display_size)) and size > 0:
+                field_params["max_length"] = size
+            elif not connection.features.supports_unlimited_varchar:
+                field_params["max_length"] = 50
+                field_notes.append(
+                    "max_length has been guessed, as the "
+                    "database returned a negative internal size and does not "
+                    "support unlimited VARCHAR"
+                )
```
IMO, we shouldn't be guessing and it's definitely no need to change that in this PR.

felixxm commented on docs/ref/models/fields.txt:
```diff
@@ -613,6 +613,11 @@ The default form widget for this field is a :class:`~django.forms.TextInput`.
+    .. note::
+
+        If you are using a database backend that supports unlimited ``varchar``
+        columns such as PostgreSQL, then you may omit this parameter.
```
`versionchanged` annotation is missing.

felixxm commented on docs/releases/4.2.txt:
```diff
@@ -140,6 +140,16 @@ Minor features
+Database backends
+~~~~~~~~~~~~~~~~~
+
+* Third-party database backends can now use callables as values for the
+  ``DatabaseWrapper.data_types`` attribute, the callable should accept a single
+  parameter which is a dictionary of the ``models.Field`` parameters.
```
This is not something that I'd document, it's rather a side-effect of the current implementation.

felixxm commented on tests/inspectdb/models.py:
```diff
@@ -106,6 +106,13 @@ class Meta:
         required_db_features = {"supports_collation_on_textfield"}
 
 
+supports_unlimited_varchar = connection.features.supports_unlimited_varchar
+
+
+class CharFieldUnlimitedVarchar(models.Model):
+    char_field = models.CharField(max_length=None if supports_unlimited_varchar else 1)
```
We should use  `Meta.required_db_features` instead:
```suggestion
class CharFieldUnlimitedVarchar(models.Model):
    char_field = models.CharField(max_length=None)

    class Meta:
        required_db_features = {"supports_unlimited_varchar"}
```

felixxm commented on tests/inspectdb/tests.py:
```diff
@@ -166,6 +166,16 @@ def test_text_field_db_collation(self):
+        out = StringIO()
+        call_command("inspectdb", "inspectdb_charfieldunlimitedvarchar", stdout=out)
+        output = out.getvalue()
+        self.assertIn(
+            "char_field = models.CharField()",
+            output,
+        )
```
This can be single-lined.

felixxm commented on django/db/backends/postgresql/base.py:
```diff
@@ -77,13 +83,13 @@ class DatabaseWrapper(BaseDatabaseWrapper):
         "DurationField": "interval",
-        "FileField": "varchar(%(max_length)s)",
-        "FilePathField": "varchar(%(max_length)s)",
+        "FileField": _get_varchar_column,
+        "FilePathField": _get_varchar_column,
```
I'm not sure if unlimited file fields are necessary, most (all?) of filesystem have natural filename length limits. I'd revert this.
```suggestion
        "FileField": "varchar(%(max_length)s)",
        "FilePathField": "varchar(%(max_length)s)",
```

Elkasitu replied:
Personally I think it's not the database's place to set these kind of restrictions, I would rather have the database not care and simply use unlimited varchars for everything and if you need to restrict the size do it higher up the stack with forms, serializers, etc. 

That said you can still set a database-level limit if you so desire, but then if the max length changes in the future you have to deal with migrations and potential table rewrites.

If you really think it should be kept as-is I'll revert.

felixxm replied:
Personally, I think this ticket should be limited to the `CharField`. I've already presented my argument against unlimited `FileField`, `FilePathField`, and `SlugField`. Django doesn't need to have builtin support for all edge cases, folks can always add custom field types in their code.

felixxm commented on django/db/backends/postgresql/base.py:
```diff
@@ -94,7 +100,7 @@ class DatabaseWrapper(BaseDatabaseWrapper):
         "PositiveIntegerField": "integer",
         "PositiveSmallIntegerField": "smallint",
-        "SlugField": "varchar(%(max_length)s)",
+        "SlugField": _get_varchar_column,
```
As far as I'm aware, slug should be short and descriptive. I don't think that providing an option for unlimited slug is a step forward.

Elkasitu replied:
anecdotally I've seen quite a few crazy slugs (but that's probably the exception rather than the norm), but I think the same argument as above applies here.

felixxm reviewed (commented):
@Elkasitu Thanks for this patch :+1: Please check my comments.

felixxm commented:
Also, a release note in the `Models` section is missing.

Elkasitu commented:
@felixxm I believe I've addressed all of your concerns now, there's one test failing but it seems like it fails randomly (previously it had failed on 3.11 and passed on 3.8)

felixxm commented on django/db/models/fields/__init__.py:
```diff
@@ -794,7 +794,11 @@ def db_type(self, connection):
         try:
-            return connection.data_types[self.get_internal_type()] % data
+            column_type = connection.data_types[self.get_internal_type()]
+            # column_type is either a single-parameter function or a string
+            if callable(column_type):
+                return column_type(data)
+            return column_type % data
```
Can we move this to `else`?
```python
try:
    column_type = connection.data_types[self.get_internal_type()]
except KeyError:
    return None
else:
    if callable(column_type):
        return column_type(data)
    return column_type % data
```

Elkasitu replied:
Sure, makes sense

felixxm reviewed (commented):
@Elkasitu Thanks :+1: 

@charettes Do you want to take a look?

felixxm reviewed (commented):
@Elkasitu Thanks for updates :+1: 

I pushed small edits and fixed `CharField.description`.

felixxm commented on django/core/management/commands/inspectdb.py:
```diff
@@ -329,7 +329,8 @@ def get_field_type(self, connection, table_name, row):
         # Add max_length for all CharFields.
         if field_type == "CharField" and row.display_size:
-            field_params["max_length"] = int(row.display_size)
+            if (size := int(row.display_size)) and size > 0:
+                field_params["max_length"] = size
```
OK, it's `-1` on `psycopg2` :disappointed: Restored.
